sdk/include/opentelemetry/sdk/metrics/controller.h

```diff
@@ -0,0 +1,150 @@
+#pragma once
+
+#include <atomic>
+#include <iostream>
+#include <sstream>
+#include <thread>
+#include <vector>
+#include "opentelemetry/exporters/ostream/metrics_exporter.h"
+#include "opentelemetry/metrics/instrument.h"
+#include "opentelemetry/nostd/unique_ptr.h"
+#include "opentelemetry/sdk/metrics/exporter.h"
+#include "opentelemetry/sdk/metrics/meter.h"
+#include "opentelemetry/sdk/metrics/processor.h"
+#include "opentelemetry/sdk/metrics/record.h"
+#include "opentelemetry/version.h"
+
+namespace metrics_api = opentelemetry::metrics;
+namespace trace_api   = opentelemetry::trace;
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace metrics
+{
+
+class PushController
+{
+
+public:
+  PushController(nostd::shared_ptr<metrics_api::Meter> meter,
+                 nostd::unique_ptr<MetricsExporter> exporter,
+                 nostd::shared_ptr<MetricsProcessor> processor,
+                 double period,
+                 int timeout = 30)
+  {
+    meter_     = meter;
+    exporter_  = std::move(exporter);
+    processor_ = processor;
+    timeout_   = (unsigned int)(timeout * 1000000);  // convert seconds to microseconds
+    period_    = (unsigned int)(period * 1000000);
+  }
+
+  /*
+   * Used to check if the metrics pipeline is currecntly active
+   *
+   * @param none
+   * @return true when active, false when on standby
+   */
+  bool isActive() { return active_.load(); }
+
+  /*
+   * Begins the data processing and export pipeline.  The function first ensures that the pipeline
+   * is not already running.  If not, it begins and detaches a new thread for the Controller's run
+   * function which periodically polls the instruments for their data.
+   *
+   * @param none
+   * @return a boolean which is true when the pipeline is successfully started and false when
+   * already active
+   */
+  bool start()
+  {
+    if (!active_.load())
```

Comment:
@reyang @ankit-bhargava - apologies, I missed this review. These checks are NOT thread-safe. I noticed this while pulling from the master. All atomic checks must be done using `atomic_exchange`. `isActive` method is not thread-safe either.

Reply:
I'm not sure exchange would be the right solution here since I don't automatically want to set `active_` to true when performing this check.  If I used exchange here, then any two consecutive calls to `start()` would mean that at least one of those passes the if statement.  In reality, `active_` should only be set to false by the stop function.  Exchange definitely make sense in the `stop` function, but if you could clarify why it is the answer in the `start` function I would appreciate it.